javascript/MaterialXView/source/viewer.js

```diff
@@ -360,33 +461,135 @@ export class Material
+                            newAssignment = new MaterialAssign(shader, collection);
+                        }
+                        else if (geom)
+                        {
+                            newAssignment = new MaterialAssign(shader, geom);
+                        }
+                        else
+                        {
+                            newAssignment = new MaterialAssign(shader, NO_GEOMETRY_SPECIFIER);
+                        }
 
+                        if (newAssignment)
+                        {
+                            this._materials.push(newAssignment);
+                        }
+                    }
+                }
+            }
+        }
+        else
+        {
+            // Search for any surface shader. It
+            // is assumed to be assigned to the entire scene
+            // The identifier used is "*" to mean the entire scene. 
+            let elem = mx.findRenderableElement(doc);
+            if (elem)
+            {
+                this._materials.push(new MaterialAssign(elem, ALL_GEOMETRY_SPECIFIER));
+            }
+        }
+        
         // Load lighting setup into document
         doc.importLibrary(viewer.getLightRig());
 
-        // Create a new material
-        const currentMaterial = viewer.getMaterial().generateMaterial(elem, viewer);
-        if (currentMaterial)
+        // Create a new shader for each material node.
+        // Only create the shader once even if assigned more than once.
+        let shaderMap = new Map();
+        let assigned = 0;
+        for (let matassign of this._materials)
         {
-            viewer.getScene().updateMaterial(currentMaterial);
+            let materialName = matassign.getMaterial().getName();
+            let shader = shaderMap[materialName];
+            if (!shader)
+            {
+                shader = viewer.getMaterial().generateMaterial(matassign.getMaterial(), viewer, searchPath);
+                shaderMap[materialName] = shader;
+            }
+            matassign.setShader(shader);
+        }
+
+        // Update scene shader assignments
+        this.updateMaterialAssignments(viewer);
+    }
+
+    //
+    // Update the assignments for scene objects based on the
+    // material assignment information stored in the viewer.
+    // Note: If none of the MaterialX assignments match the geometry
+    // in the scene, then the first material assignment shader is assigned
+    // to the entire scene.
+    //
+    updateMaterialAssignments(viewer)
+    {
+        let assigned = 0;
+        for (let matassign of this._materials)
+        {
+            if (matassign.getShader())
+            {
+                assigned += viewer.getScene().updateMaterial(matassign);
+            }
+        }
+        if (assigned == 0 && this._materials.length)
+        {
+            this._defaultMaterial = new MaterialAssign(this._materials[0].getMaterial(), ALL_GEOMETRY_SPECIFIER);
```

Comment:
Assign to first material if no material assignments match geometry. Similar to desktop except desktop has concept of "active" material which does not exist here.